OpenRA.Mods.Common/Activities/HarvesterDockSequence.cs

```diff
@@ -100,16 +112,35 @@ public override bool Tick(Actor self)
 
 		public override IEnumerable<Target> GetTargets(Actor self)
 		{
-			yield return Target.FromActor(Refinery);
+			yield return Target.FromActor(RefineryActor);
 		}
 
 		public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
 		{
-			yield return new TargetLineNode(Target.FromActor(Refinery), Color.Green);
+			yield return new TargetLineNode(Target.FromActor(RefineryActor), Color.Green);
 		}
 
 		public abstract void OnStateDock(Actor self);
 
 		public abstract void OnStateUndock(Actor self);
+
+		void NotifyDocked(Actor self)
+		{
+			foreach (var nd in notifyDockClients)
+				nd.Docked(self, RefineryActor);
+
+			foreach (var nd in notifyDockHosts)
+				nd.Docked(RefineryActor, self);
+		}
+
+		void NotifyUndocked(Actor self)
+		{
+			foreach (var nd in notifyDockClients)
+				nd.Undocked(self, RefineryActor);
+
+			if (RefineryActor.IsInWorld && !RefineryActor.IsDead)
```

Comment:
It doesn't make sense that this is the only `foreach` of the 4 basically identical ones that has this check. This `if` should be present at the start of both notification methods and bail them out early.
But since the exact same `if` is already present at lease once in the `switch` in `Tick` (and missing in at least 2 cases there!), this probably needs a better generalisation (check in `Tick`?)

Reply:
Clients don't need checks, and the NotifyDocked checks are done in Tick